lib/astutils.cpp

```diff
@@ -1023,6 +1023,86 @@ static bool isZeroBoundCond(const Token * const cond)
     return false;
 }
 
+static bool isCastToBool(const Token * const tok)
+{
+    if (isCPPCast(tok)) {
+        if (tok->astOperand1()->next()->next()->str() == "bool")
+            return true;
+    } else if (tok->astOperand1()->str() == "bool")
+        return true;
+    return false;
+}
+
+static bool isForLoopCondition(const Token * const tok)
+{
+    if (!tok || !tok->astParent())
+        return false;
+    const Token *const parent = tok->astParent();
+    return parent->str() == ";" && parent->astOperand1() == tok &&
+           parent->astParent() && parent->astParent()->str() == ";" &&
+           parent->astParent()->astParent() && parent->astParent()->astParent()->str() == "(" &&
+           parent->astParent()->astParent()->astOperand1()->str() == "for";
+}
+
+/**
+ * Is token used a boolean (cast to a bool, or used as a condition somewhere)
+ * @param tok
+ * @param checkingParent true if we are checking a parent. This is used to know
+ * what we are checking. For instance in `if (i == 2)`, isUsedAsBool("==") is
+ * true whereas isUsedAsBool("i") is false, but it might call
+ * isUsedAsBool_internal("==") which must not return true
+ */
+static bool isUsedAsBool_internal(const Token * const tok, bool checkingParent)
+{
+    if (!tok)
+        return false;
+    const Token::Type type = tok->tokType();
+    if (type == Token::eBitOp || type == Token::eIncDecOp || (type == Token::eArithmeticalOp && (tok->str() != "*" || tok->astOperand2())))
+        // those operators don't return a bool
+        return false;
+    if (type == Token::eComparisonOp) {
+        if (!checkingParent)
+            // this operator returns a bool
+            return true;
+        if (Token::Match(tok, "==|!="))
+            return astIsBool(tok->astOperand1()) || astIsBool(tok->astOperand2());
+        return false;
+    }
+    if (type == Token::eLogicalOp)
+        return true;
+    if (astIsBool(tok))
+        return true;
+
+    const Token * const parent = tok->astParent();
+    if (!parent)
+        return false;
+    if (parent->str() == "(" && parent->astOperand2() == tok) {
+        if (Token::Match(parent->astOperand1(), "if|while"))
+            return true;
+
+        if (parent->isCast()) {
+            if (isCastToBool(parent))
```

Comment:
Couldn't this be `return astIsBool(parent)`? Or is the valueType not available for casts?

Reply:
Indeed, isCastToBool is useless. However, we don't want to return, as if we could still have a conversion to bool after the cast, for instance `!static_cast<int>(i)`. I've clean up the code.